project/Build.scala

```diff
@@ -675,6 +675,94 @@ object Build {
     recur(lines)
   }
 
+  val shadedSourceGenerator = (Compile / sourceGenerators) += Def.task {
+    val downloads = Seq(
+      "https://repo1.maven.org/maven2/com/lihaoyi/pprint_3/0.9.3/pprint_3-0.9.3-sources.jar",
+      "https://repo1.maven.org/maven2/com/lihaoyi/fansi_3/0.5.1/fansi_3-0.5.1-sources.jar",
+      "https://repo1.maven.org/maven2/com/lihaoyi/sourcecode_3/0.4.3-M5/sourcecode_3-0.4.3-M5-sources.jar",
```

Comment:
...wait, is 0.4.3-M5 a stable version? 🤔

Reply:
if it's become stabilised, by all means. 👍 
I'd rather avoid milestone versions here.

Reply:
@hamzaremmal we're basically shading the library here. The advantage of this approach is that it won't collide with user code: if I'm using a project with a different version of `sourcecode`, `pprint`, or `fansi`, I don't want the classpath to have multiple jars containing the same classfiles which can result in a non-deterministic version of each classfile being loaded when someone opens the Scala console.

The builtin Scala `repl` does not appear to have any classloader isolation to keep the dotty internal classfiles separate from the user-provide classpath and commands, and so we should try to keep the dotty internal classfiles as encapsulated as possible

Reply:
Also, others have raised concerns in the pat about using Scala libraries in the compiler codebase affecting the bootstrapping process. By building from source, we treat it effectively as Dotty's own source files, removing any divergence in the code paths: they are treated identically to scala3's own sources. If scala3 can compile itself, it should be able to compile these sources without issue

Reply:
> We should indeed compile these from sources. We can depend on binaries for Java libraries (hence jline and asm are fine), but not for Scala libraries.

Could you explain why? I thought Scala is maintaining backwards binary/tasty compatibility. Doesn't that mean we shohld always be able to depend on older scala 3 jars in the scala3 compiler regardless of how kuch bootstrapping we do?

Reply:
Because circular dependencies are evil. They're not *too* evil across time (Av2 -> Bv1 -> Av1), but they're still difficult to reason about.

And even though Scala 3 will forever be backward compat, an eventual Scala 4 wouldn't. We shouldn't paint our build into a corner. Scala 2 tried this several times over its lifetime, and rolled back every time. It's a massive pain every time it happens. There would need to be a huge upside to depending on a binary for that to be offset.